src/StreamJsonRpc/JsonRpc.cs

```diff
@@ -1553,10 +1716,63 @@ private void CancelPendingOutboundRequest(object state)
             }).Forget();
         }
 
+        private void TraceLocalMethodAdded(string rpcMethodName, MethodSignatureAndTarget targetMethod)
+        {
+            Requires.NotNullOrEmpty(rpcMethodName, nameof(rpcMethodName));
+
+            if (this.TraceSource.Switch.ShouldTrace(TraceEventType.Information))
+            {
+                this.TraceSource.TraceEvent(TraceEventType.Information, (int)TraceEvents.LocalMethodAdded, "Added local RPC method \"{0}\" -> {1}", rpcMethodName, targetMethod);
+            }
+        }
+
+        private void TraceMessageSent(JsonRpcMessage message)
+        {
+            if (this.TraceSource.Switch.ShouldTrace(TraceEventType.Information))
+            {
+                this.TraceSource.TraceData(TraceEventType.Information, (int)TraceEvents.MessageSent, message);
+            }
+
+            if (this.TraceSource.Switch.ShouldTrace(TraceEventType.Verbose))
+            {
+                this.TraceSource.TraceData(TraceEventType.Verbose, (int)TraceEvents.MessageSent, "Sent: {0}", this.GetMessageJson(message));
+            }
+        }
+
+        private void TraceMessageReceived(JsonRpcMessage message)
+        {
+            if (this.TraceSource.Switch.ShouldTrace(TraceEventType.Information))
+            {
+                this.TraceSource.TraceData(TraceEventType.Information, (int)TraceEvents.MessageReceived, message);
+            }
+
+            if (this.TraceSource.Switch.ShouldTrace(TraceEventType.Verbose))
+            {
+                this.TraceSource.TraceData(TraceEventType.Verbose, (int)TraceEvents.MessageReceived, "Received: {0}", this.GetMessageJson(message));
+            }
+        }
+
+        private object GetMessageJson(JsonRpcMessage message)
+        {
+            try
+            {
+                return JToken.FromObject(message);
+            }
+            catch (Exception ex)
+            {
+                // In the event that a formatter is in use that is not JsonMessageFormatter,
+                // there is no guarantee that the message is serializable with Newtonsoft.Json.
+                // This was a best effort, and it didn't work out.
+                return $"<JSON representation not available: {ex.Message}>";
```

Comment:
Should there be a method on `IJsonRpcMessageFormatter` to get a string representation of a message for tracing?

Reply:
I was thinking about that too... Maybe. But then I'd also have to add a `IJsonRpcMessageHandler.Formatter` property so that I could retrieve the formatter. 
Given you had the same idea, maybe I'll go ahead and do that.

Reply:
Done.